File: ruby/meetup/meetup.rb
require 'minitest/autorun'
require 'date'

class Meetup

  WEEKDAYS = {
    sunday:    0,
    monday:    1,
    tuesday:   2,
    wednesday: 3,
    thursday:  4,
    friday:    5,
    saturday:  6,
  }

  def initialize(month, year)
    @month, @year = month, year
  end

  def day(weekday, schedule)
    weekday_number = WEEKDAYS.fetch(weekday)
    case schedule
    when :first  then first(weekday_number)
    when :last   then last(weekday_number)
    when :second then nth(1, weekday_number)
    when :third  then nth(2, weekday_number)
    when :fourth then nth(3, weekday_number)
    when :teenth then teenth(weekday_number)
    end
  end

  private

  def first(weekday)
    advance_to_first(date_in_month(1), weekday)
  end

  def last(weekday)
    advance_to_first(date_in_month(-7), weekday)
  end

  def nth(week, weekday)
    advance_by_weeks(first(weekday), week)
  end

  def teenth(weekday)
    advance_to_first(date_in_month(13), weekday)
  end

  def advance_to_first(date, weekday)
    date + (weekday - date.wday) % 7
  end

  def advance_by_weeks(date, number_of_weeks)
    date + number_of_weeks * 7
  end

  def date_in_month(day_of_month)
    Date.civil(@year, @month, day_of_month)
  end

end

File: elixir/meetup/meetup.exs
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
      :monday | :tuesday | :wednesday
    | :thursday | :friday | :saturday | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date

  def meetup(year, month, weekday, :first) do
    advance_to_first({year, month, 1}, weekday)
  end

  def meetup(year, month, weekday, :second) do
    advance_to_first({year, month, 1}, weekday) |> advance_date_by_weeks(1)
  end

  def meetup(year, month, weekday, :third) do
    advance_to_first({year, month, 1}, weekday) |> advance_date_by_weeks(2)
  end

  def meetup(year, month, weekday, :fourth) do
    advance_to_first({year, month, 1}, weekday) |> advance_date_by_weeks(3)
  end

  def meetup(year, month, weekday, :last) do
    advance_to_first({year, month, last_day_of_month(year, month) - 6}, weekday)
  end

  def meetup(year, month, weekday, :teenth) do
    advance_to_first({year, month, 13}, weekday)
  end

  defp advance_date_by_weeks({year, month, day}, weeks) do
    {year, month, day + weeks * 7}
  end

  defp advance_to_first(date, weekday) do
    from_weekday   = weekday_number(weekday)
    to_weekday     = date_weekday_number(date)
    offset_in_days = weekdays_between(from_weekday, to_weekday)
    if offset_in_days == 0 do
      date
    else
      date
      |> date_to_epoch
      |> advance_epoch_by_days(offset_in_days)
      |> epoch_to_date
    end
  end

  @day_seconds 24 * 60 * 60

  defp advance_epoch_by_days(epoch, days) do
    epoch + days * @day_seconds
  end

  defp date_to_epoch(date) do
    :calendar.datetime_to_gregorian_seconds({date, {0, 0, 0}})
  end

  defp epoch_to_date(epoch) do
    {date, _} = :calendar.gregorian_seconds_to_datetime(epoch)
    date
  end

  defp weekdays_between(from_weekday, to_weekday) do
    rem(7 - (to_weekday - from_weekday), 7)
  end

  defp last_day_of_month(year, month) do
    :calendar.last_day_of_the_month(year, month)
  end

  defp date_weekday_number(date) do
    :calendar.day_of_the_week(date)
  end

  defp weekday_number(:monday),    do: 1
  defp weekday_number(:tuesday),   do: 2
  defp weekday_number(:wednesday), do: 3
  defp weekday_number(:thursday),  do: 4
  defp weekday_number(:friday),    do: 5
  defp weekday_number(:saturday),  do: 6
  defp weekday_number(:sunday),    do: 7

end
